### PR TITLE
Counting Haybales Solution

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

i can add a short explanation if needed. this would probably be fine as a user sol as well, but idk

### CODE REVIEW

File: solutions/platinum/usaco-578.mdx [hunk 1]
```diff
@@ -0,0 +1,126 @@
+---
+id: usaco-578
+source: USACO Platinum 2015 December
+title: Counting Haybales
+author: Justin Ji
+---
+
+[Official Analysis](https://usaco.org/index.php?page=viewproblem2&cpid=578)
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-578.mdx:8]: is this link uh
correct

File: solutions/platinum/usaco-578.mdx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	inline array<ll, 2> join(const array<ll, 2> &a, const array<ll, 2> &b) {
+		return {a[0] + b[0], min(a[1], b[1])};
+	}
+
+	LazySegtree(const vector<int> &a)
+	    : sz((int)a.size()), t(4 * sz), lz(4 * sz) {
+		build(1, 0, sz - 1, a);
+	}
+
+	void build(int v, int tl, int tr, const vector<int> &a) {
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-578.mdx:39]: why not put directly in constructor

**Review by SansPapyrus683 — APPROVED** [review]
is this just a template problem what

**Comment by TheGamingMousse** [comment]
> is this just a template problem what

lol it basically is 💀

File: solutions/platinum/usaco-578.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-578.mdx:39]: like a recursive lambda or smth?

Comment (SansPapyrus683) [reply solutions/platinum/usaco-578.mdx:39]: can you make this and other functions that aren't directly called private

File: solutions/platinum/usaco-578.mdx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	LazySegtree(const vector<int> &a)
+	    : sz((int)a.size()), t(4 * sz), lz(4 * sz) {
+		build(1, 0, sz - 1, a);
+	}
+
+	/** @return {sum, minimum value} of haybales in [ql, qr] */
+	array<ll, 2> qry(int ql, int qr) { return qry(1, 0, sz - 1, ql, qr); }
+
+	/** adds to every haybale in the range [ql, qr] */
+	void upd(int ql, int qr, int amt) { upd(1, 0, sz - 1, ql, qr, amt); }
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-578.mdx:99]: could you mayhaps rename these two to `range_info` and `inc_range`

Comment (TheGamingMousse) [reply solutions/platinum/usaco-578.mdx:99]: sounds good, ima do that in a sec

File: solutions/platinum/usaco-578.mdx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+int main() {
+	freopen("haybales.in", "r", stdin);
+	int n, q;
+	cin >> n >> q;
+	vector<int> a(n);
+	for (int &i : a) { cin >> i; }
+
+	LazySegtree st(a);
+	freopen("haybales.out", "w", stdout);
```
Comment (ryanchou-dev) [inline solutions/platinum/usaco-578.mdx:114]: i would prefer to keep the freopens together, but it's totally optional

**Review by ryanchou-dev — APPROVED** [review]
fire problem

**Pull request merged**